### PR TITLE
add step in allure report for spock framework

### PR DESCRIPTION
**Pull request by alexey-shnaydmiller**
#### Checklist
- [x] [Sign Allure CLA][cla]
- [x] Provide unit tests

[cla]: https://cla-assistant.io/accept/allure-framework/allure2

### CODE REVIEW

File: allure-spock/src/main/java/io/qameta/allure/spock/AllureSpock.java [hunk 1]
```diff
@@ -308,4 +329,49 @@ public void afterIteration(final IterationInfo iteration) {
...
+        }
+        tempFeature.setIterationNameProvider(new UnrollNameProvider(tempFeature, input));
+        return tempFeature.getIterationNameProvider().getName(iteration);
+    }
+
+    private boolean isEmptyOrContainsOnlyEmptyStrings(final List<String> strings) {
+        final boolean countOnlyEmptyStrings = strings
+                .stream()
+                .noneMatch(StringUtils::isNotBlank);
+        return CollectionUtils.isEmpty(strings) || countOnlyEmptyStrings;
```
Comment (baev) [inline allure-spock/src/main/java/io/qameta/allure/spock/AllureSpock.java:375]: you introduce dependency only for one method? Why not to use `Objects.isNull(strings) || strings.isEmpty() || ...` instead?

Comment (alexey-shnaydmiller) [reply allure-spock/src/main/java/io/qameta/allure/spock/AllureSpock.java:375]: Yes, I made an dependency by discussing with Eroshenko

File: allure-spock/build.gradle.kts [hunk 2]
```diff
@@ -5,11 +5,15 @@ apply(plugin = "groovy")
 val agent: Configuration by configurations.creating
 
 val spockFrameworkVersion = "1.2-groovy-2.5"
+val apacheCommonsCollection4 = "4.4"
```
[the remark this answers is not in the document]
Comment (alexey-shnaydmiller) [reply allure-spock/build.gradle.kts:8]: Need to specify dependency without version or in the main?

File: allure-spock/src/main/java/io/qameta/allure/spock/AllureSpock.java [hunk 3]
```diff
@@ -28,24 +28,17 @@
 import org.junit.runner.Description;
 import org.spockframework.runtime.AbstractRunListener;
 import org.spockframework.runtime.extension.IGlobalExtension;
-import org.spockframework.runtime.model.ErrorInfo;
-import org.spockframework.runtime.model.FeatureInfo;
-import org.spockframework.runtime.model.IterationInfo;
-import org.spockframework.runtime.model.SpecInfo;
+import org.spockframework.runtime.extension.builtin.UnrollNameProvider;
+import org.spockframework.runtime.model.*;
```
[the remark this answers is not in the document]
Comment (alexey-shnaydmiller) [reply allure-spock/src/main/java/io/qameta/allure/spock/AllureSpock.java:32]: 5 и 6 build show.

**Comment by baev** [comment]
as far as I get from test you provided -- framework blocks are always displayed before all the steps (from allure)?

**Comment by alexey-shnaydmiller** [comment]
The test shows that the allure report prints all the blocks from the spock. Depending on which block of the spock is used, it is printed in the report.

**Comment by baev** [comment]
I mean if you have test like this:

```groovy
given: "given"
step "given step"
when: "when"
step "when step"
then: "then"
step "then step"
```

you'll see the following scenario in the report:

```
"given"
"when"
"then"
"given step"
"when step"
"then step"
```

instead of correct nesting

```
"given"
    "given step"
"when"
    "when step"
"then"
    "then step"
```

or flat 

```
"given"
"given step"
"when"
"when step"
"then"
"then step"
```

**Comment by alexey-shnaydmiller** [comment]
I get it. It was just done not to write a Step. I'll fix that shortcoming tomorrow.

**Comment by baev** [comment]
that is the main concern about this pull request. A wish to find a way to emit events when block start in order to support steps and attachments within blocks. Otherwise we'll immediately get an issue complaining about that.

Overall, I see that this one did improve integration for spock users in some way, so I'm down to merge it

**Comment by baev** [comment]
closed via #838

**Pull request closed without merging**